src/pudl/analysis/plant_parts_eia.py

```diff
@@ -1334,56 +1034,83 @@ def add_record_count_per_plant(self, part_df: pd.DataFrame) -> pd.DataFrame:
-        bool_df = true_grans[
-            self.id_cols
-            + IDX_TO_ADD
-            + IDX_OWN_TO_ADD
-            + [
-                f"true_gran_{self.part_name}",
-                f"appro_part_label_{self.part_name}",
-                f"record_id_eia_{self.part_name}",
-            ]
-        ].drop_duplicates()
-
-        prop_true_len1 = len(bool_df[bool_df[f"true_gran_{self.part_name}"]]) / len(
-            bool_df
-        )
-        logger.debug(f"proportion of trues: {prop_true_len1:.02}")
-        logger.debug(f"number of records pre-merge:  {len(part_df)}")
+        First the plant part list records are matched to generators. Then
+        the matched records are sorted by PLANT_PARTS_ORDERED and the
+        highest granularity record for each generator is marked as the true
+        granularity. The appropriate true granular part label and record id
+        is then merged on to get the plant part table with true granularity labels.
 
-        part_df = part_df.merge(
-            bool_df,
-            on=self.id_cols + IDX_TO_ADD + IDX_OWN_TO_ADD,
-            how="left",
-        ).rename(
+        Arguments:
+            ppl: (pd.DataFrame) The plant parts list
+        """
+        parts_to_gens = match_to_single_plant_part(
+            multi_gran_df=ppl,
+            ppl=ppl,
+            part_name="plant_gen",
+            cols_to_keep=["plant_part"],
+        )[["record_id_eia_og", "record_id_eia", "plant_part_og"]].rename(
             columns={
-                f"true_gran_{self.part_name}": "true_gran",
-                f"appro_part_label_{self.part_name}": "appro_part_label",
-                f"record_id_eia_{self.part_name}": "appro_record_id_eia",
+                "record_id_eia": "gen_id",
+                "record_id_eia_og": "record_id_eia",
+                "plant_part_og": "plant_part",
             }
         )
+        # concatenate the gen id's to get the combo of gens for each record
+        combos = (
+            parts_to_gens.groupby(["record_id_eia"])["gen_id"]
+            .apply(lambda x: ",".join(x))
+            .rename("gens_combo")
+        )
+        parts_to_gens = parts_to_gens.merge(
+            combos, how="left", left_on="record_id_eia", right_index=True
+        )
 
-        prop_true_len2 = len(part_df[part_df.true_gran]) / len(part_df)
-        logger.debug(f"proportion of trues: {prop_true_len2:.02}")
-        logger.debug(f"number of records post-merge: {len(part_df)}")
-        return part_df
+        # categorical columns allow sorting by PLANT_PARTS_ORDERED
+        parts_to_gens["plant_part"] = pd.Categorical(
+            parts_to_gens["plant_part"], PLANT_PARTS_ORDERED
+        )
+        parts_to_gens = parts_to_gens.sort_values("plant_part")
+        # get the true gran records by finding duplicate gen combos
+        dupes = parts_to_gens.duplicated(subset=["gens_combo"], keep="first")
+        parts_to_gens.loc[:, "true_gran"] = ~(dupes)
```

Comment:
do we need to add in the records that are truly unique? like the ones that don't have any other `record_id_eia`'s that share the composite `gen_id`'s? Or more generally how are the non-duplicated granularities treated in here

Reply:
I'm pretty sure non-duplicated granularities are treated correctly here (they are marked as true granularities). Here's the logic:
- `dupes` (generated from the `duplicated`) marks duplicates as `True` except for the first occurrence, so non-duplicated granularities will be marked as `False`
- When assigning the `true_gran` column to `~dupes` these `False` non-duplicated granularities will be marked as `True`. 

This should be correct? In the unit test that I made (which corresponds to the plant 3 example you wrote up in the google sheet) there is a non-duplicated granularity in there - `unit_b` corresponds to `gen_1` and `gen_2`, which is a non duplicated granularity. `unit_b` is then marked as a true granularity.

Reply:
ah yes you are so right! ty for explaining. this makes sense now... I think i was having fuzzy brain when I was looking at this. It might be good to add this logic as inline comment!

Reply:
added this as a comment!